content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+To visualize what happens when this is not the case, let’s take a look at the demo application that Dan Abramov presented at his talk, [Beyond React 16](https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html), at JSConf Iceland 2018.
+
+The application works like this: The more you type into the input below, the more detailed the charts below will get. Since both of the updates (the input element and the chart) run at the same time, the browser has to do so much computation that it will drop some frames. This leads to noticeable delays and a bad user experience:
+
+<video src="/blog/scheduling-in-react/sync-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+However, in this example, it’s more important to update the input with the new text than it is to update the charts. A version that prioritizes the input will appear a lot more responsive to the end user even though the same computation time is required:
+
+<video src="/blog/scheduling-in-react/concurrent-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+Unfortunately, current user interface architectures makes it non-trivial to implement this kind of prioritization. One way to solve this is by [debouncing](https://davidwalsh.name/javascript-debounce-function) the chart update. The problem with this is that the charts still render synchronously when the debounced callback fires, which will again cause our user interface to take multiple seconds during which it is not responsive. We can do better!
+
+## Browser Event Loop
+
+Before we learn more about how this can be achieved, let’s dig deeper and understand why the browser has issues with these kind of user interactions.
+
+JavaScript code is executed in one thread, meaning that only one line of JavaScript can be run at any given time. The same thread is also responsible for other document lifecycles like layout and paint.[^1] This means that whenever JavaScript code runs, the browser is blocked from doing anything else.
+
+To keep the user interface responsive, we only have a very short timeframe before we need to be able to receive the next input events. In the browser run loop visualization presented in Subhie Panicker’s and Jason Miller’s talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread), at the Chrome Dev Summit 2018 below, we can see that we only have 16 milliseconds (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:
+
+![The browser event loop starts by running input handlers, followed by animation frame callbacks, and ends with document lifecycles (style, layout, paint). All of this should complete within one frame which is approximately 16 milliseconds on a 60Hz display.](event-loop-browser.png)
```

Comment:
Hmm, not sure if this should be "that" or "which" and not sure how to properly communicate the difference.

"All of this should complete within one frame which is approximately 16 milliseconds on a 60Hz display."

Basically, it's

"All of this should complete within one frame, which is approximately 16 milliseconds on a 60Hz display."

vs.

"All of this should complete within one frame that is approximately 16 milliseconds on a 60Hz display."

The first means more there is only one frame, and it will be approx 16 ms, but that information is a "given." The second means more like there could be multiple frames but the one we're referring to is the one that's approx 16 ms as opposed to one that could be different. Confusing, sorry. Let me know if I should try explaining better.

Reply:
No totally understood you there! It's the first one; one frame is usually (expect in some rare case) referring to the time of 16 milliseconds since most displays refresh at 60 times per seconds. 🙂